### ISSUE TITLE
Possible false positive with UNT0008 when nullable type is used

### ISSUE THREAD

**Issue by NoTuxNoBux** [microsoft/Microsoft.Unity.Analyzers#235]
#### Bug description

- Version of analyzers assembly: 1.13.0
- Analyzer rule: [UNT0008](https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0008.md)
- Error (exception message, type, and callstack where applicable): 

```
Unity objects should not use null propagation.
```

#### To Reproduce

```cs
class A
{
    private GameObject foo;
    private GameObject? bar;

    public void Method()
    {
        foo?.SetActive(true); // Reports correctly.
        bar?.SetActive(true); // Reports incorrectly.
    }
}
```

#### Expected behavior
In the case of explicitly nullable types, null propagation can be used validly to check if the object is actually (and only) `null`, much like `is null` and `is not null`, all of which disregard the equality operator (intentionally).

#### Additional context
I'm not entirely sure what the desired behaviour is with explicitly nullable types. It's still a risk to use null propagation because of the equality operator, but in some cases this may be intentional (e.g. initialize-once variable, which doesn't care if the underlying object is destroyed or not). At some point, though, explicitly nullable types will become the default, and if the warning is disabled in this scenario, people might accidentally be using it.

Still, though, in a scenario where everything migrated to explicitly nullable types, why would you be comparing a non-nullable GameObject through `== null`? You will want to make your intent explicit: if it can be _truly_ `null`, you have to add the question mark, and both `is null` or `== null` could be intentional. If you don't want to make it explicitly nullable, `is null` and null propagation make little sense as they don't work at all (they skip the operator overload), but you can still do `== null` because of Unity's overload, which would be weird to see - in this scenario it would probably best if Unity replaced the operator overload with `GameObject.IsDestroyed` or something instead, which makes it implicit that the instance still exists.

**Comment by sailro** [comment]
cc @jbevain

**Comment by tannergooding** [comment]
Was going to come here and open an issue as well.

`?.` in C# does not use `operator ==`. It is equivalent to using `is null` or `(object)x == null` and does a pure null check ignoring any user defined semantics.

It is therefore safe to use since it is purely a defense against dereferencing a null.

You can trivially see this in an example such as https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs4fsNKTuAQIuOhAAOMKAEMM0OgF55ACha4AFgGc0zOlE0BKBQD46AMym4NMANwcudzrQZ8BQkeMky5AQmWrN2ix6GobyJuaWNg500U68/LwADHQAKjAaGCp0YKEmYAD8AHQpEADKGFDYAHYA5kr6tpQAvkA===

Where the IL for `static string Test(C c) => c?.ToString();` is:
```
    .method public hidebysig static 
        string Test (
            class C c
        ) cil managed 
    {
        // Method begins at RVA 0x20a0
        // Code size 12 (0xc)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: brtrue.s IL_0005

        IL_0003: ldnull
        IL_0004: ret

        IL_0005: ldarg.0
        IL_0006: callvirt instance string [System.Runtime]System.Object::ToString()
        IL_000b: ret
    } // end of method C::Test
```

**Comment by jbevain** [comment]
@tannergooding it is safe to use but using null propagation and other null coalescing on Unity objects is inherently ambiguous due to Unity's special handling of null comparisons for destroyed game objects. Our analyzer provides an information here (and not a warning) that you might want to refactor your code to be explicit about the intent.

**Comment by jbevain** [comment]
@NoTuxNoBux C# nullable types is also something that is going to be ambiguous with regards to Unity's special null handling. Imagine you have:

```csharp
#nullable enable

public static class Foo
{
    public static void Bar(GameObject go)
    {
        go.Destroy();
        if (go == null)
        {
            Debug.Log("GameObject is null");
        }
    }
}
```

In the method `Bar`, the C# compiler thinks that `go` can not be null. But after a call to .Destroy, `== null` is returning true. That goes pretty much against the feature. `GameObject?` is ambiguous too. I'm not sure how we should approach nullable reference types in the context of Unity.

**Comment by NoTuxNoBux** [comment]
> In the method Bar, the C# compiler thinks that go can not be null. But after a call to .Destroy, == null is returning true. That goes pretty much against the feature.

I understand what you mean, and I wanted to add some context here: Unity's `GameObject` explicitly overloads `operator==`, so even in a nullable context the compiler _could_ see that `== null` makes sense where an actual null check does not. As Unity's `operator==` overload references `Object` as type, and is not in a nullable context, the compiler could deduce that `== null` makes sense in this context - it can't not do so because whether it is depends on what the operator overload says, even if the actual variable can never be fully `null` (i.e. it can still be truthy to compare to value `null` even though the object itself is not ever `null`).
